mixnet/src/fragment.rs

```diff
@@ -1,22 +1,249 @@
+use std::{collections::HashMap, ops::Deref};
+
+use sphinx_packet::{constants::PAYLOAD_SIZE, payload::PAYLOAD_OVERHEAD_SIZE};
+use uuid::Uuid;
+
 use crate::error::MixnetError;
 
+pub(crate) struct FragmentSet(Vec<Fragment>);
+
+const MAX_PLAIN_PAYLOAD_SIZE: usize = PAYLOAD_SIZE - PAYLOAD_OVERHEAD_SIZE;
+
+impl FragmentSet {
+    pub(crate) fn new(msg: &[u8]) -> Result<Self, MixnetError> {
+        let chunk_size = MAX_PLAIN_PAYLOAD_SIZE - FRAGMENT_HEADER_SIZE;
+
+        // For now, we don't support more than max_fragments() fragments.
+        // If needed, we can devise the FragmentSet chaining to support larger messages, like Nym.
+        let num_chunks = Self::num_chunks(msg, chunk_size);
+        if num_chunks > MAX_NUM_FRAGMENTS {
+            return Err(MixnetError::MessageTooLong(msg.len()));
+        }
+
+        let set_id = Uuid::new_v4();
+        Ok(FragmentSet(
+            msg.chunks(chunk_size)
+                .enumerate()
+                .map(|(i, chunk)| Fragment {
+                    header: FragmentHeader {
+                        set_id,
+                        total_fragments: num_chunks as FragmentId,
+                        fragment_id: i as FragmentId,
+                    },
+                    body: Vec::from(chunk),
+                })
+                .collect(),
+        ))
+    }
+
+    fn num_chunks(msg: &[u8], chunk_size: usize) -> usize {
+        msg.len() / chunk_size + (msg.len() % chunk_size > 0) as usize
+    }
+}
+
+impl Deref for FragmentSet {
+    type Target = Vec<Fragment>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
```

Comment:
`Deref` is often use for smart pointers. In this cases you would want to use `AsRef` instead. Why? Because it is an **explicit** call instead of an implicit one. So you know what is going on when reading it, also it means that you are sure you actually want it to behave as that. It is up to the caller to decide, not to the user.

Reply:
Thank you. I forgot that there is `AsRef` trait :) 4ec973b